File: practice/golang/leetcode/sortLinkedList/sortLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def sortList(self, head):
        step = 1
        size = self.getListSize(head)
        while head is not None and step < size:
            head = self.orderOnce(head,step)
            step *= 2
        return head

    def orderOnce(self, head,step):
        previoustail = None
        top = None
        while head is not None:
            # reorder every step nodes and join them
            node1, node2, nexthead = self.getTwoList(head,step)
            head,tail = self.switch(node1,node2)        
            if previoustail is not None:
                previoustail.next = head
            previoustail = tail
            if top is None:
                top = head
            tail.next = nexthead       
            head = tail.next
        return top

    def getListSize(self,head):
        counter = 0
        while head is not None:
            head = head.next
            counter += 1
        return counter

 
    def getTwoList(self, head, step):
        # reorder every step nodes and join them
        node1 = head
        counter = 0
        while head is not None and counter < step-1:
            head = head.next
            counter += 1
        if head is None:
            return node1, None, None
        node2 = head.next
        head.next = None
        head = node2
        counter = 0
        while head is not None and counter < step-1:
            head = head.next
            counter += 1
        nexthead = None
        if head is not None:
            nexthead = head.next                  
            head.next = None
        return node1, node2, nexthead
 

    def switch(self, node1, node2):
        previous = None
        head = None
        if node2 is None:
            head = node1
            previous = node1
            while previous.next is not None:
                previous = previous.next
            return head, previous


        if node1.val < node2.val:
            previous = node1
            node1 = node1.next
        else:
            previous = node2
            node2 = node2.next
        head = previous
        while node1 is not None and node2 is not None:
            if node1.val < node2.val:
                previous.next = node1
                node1 = node1.next
            else:
                previous.next = node2
                node2 = node2.next
            previous = previous.next
        if node2 is not None:
            previous.next = node2
        if node1 is not None:
            previous.next = node1
        while previous.next is not None:
            previous = previous.next
        #print "after switch:", head.val, previous.val
        return head,previous

File: practice/golang/leetcode/sortLinkedList/sortLinkedList.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val int
    Next *ListNode
}

func merge(l *ListNode, r *ListNode) *ListNode {
    if l == nil {
        return r
    }
    if r == nil {
        return l
    }
    var p *ListNode
    if (l.Val < r.Val){
       p = l
       p.Next = merge(l.Next, r)
    } else {
	   p = r
       p.Next = merge(l, r.Next)
	}
    return p
}

func sortList(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
	}
    slow := head
	fast := head.Next
    for fast != nil && fast.Next != nil {
        slow = slow.Next
        fast = fast.Next.Next
	}
    left := head
    right := slow.Next
    slow.Next = nil
    left = sortList(left)
    right = sortList(right)
	return merge(left, right)
}

func formatList(input []int) *ListNode {
    if len(input) == 0 {
        return nil
	}
	var tail *ListNode
    for i := len(input)-1; i >= 0; i-- {
        cnode := &ListNode{
           Val: input[i],
           Next: nil,
        }
        if tail != nil {
            cnode.Next = tail
		}
        tail = cnode
	}
    return tail
}

func printList(head *ListNode) {
	for head != nil {
        if head.Next == nil {
            fmt.Printf("%d", head.Val)
        } else {
		    fmt.Printf("%d -> ", head.Val)
		}
        head = head.Next
	}
    fmt.Println()
}

func main(){
    head := formatList([]int{6,9,3,-1,0,2,3,4})
    printList(head)
	head = sortList(head)
    printList(head)
}
